File: scripts/prepare_output_head.py
#!/usr/bin/env python3
"""Freeze new train/dev/calibration/test texts before fitting a deployment head."""
import argparse, collections, copy, csv, io, json, random, zipfile
from pathlib import Path
from kaggle_airline import ARCHIVE_SHA256, INSTRUCTION, OPTIONS, identity, SOURCE
from kaggle_ag_news import sha256


def prepare(source, previous, output):
    archive=source/'dataset.zip'
    assert sha256(archive)==ARCHIVE_SHA256
    excluded={identity(json.loads(s)['request']['state']['tweet']) for folder,names in
              [(source,['fit','validation']),(previous,['train','calibration','test'])]
              for name in names for s in (folder/(name+'.jsonl')).read_text().splitlines()}
    with zipfile.ZipFile(archive) as z:
        rows=list(csv.DictReader(io.StringIO(z.read('Tweets.csv').decode('utf-8-sig'))))
    groups=collections.defaultdict(list)
    for i,r in enumerate(rows,1): groups[identity(r['text'])].append((i,r))
    available={k:[] for k in OPTIONS}
    for key,copies in groups.items():
        if key and key not in excluded and len({r['airline_sentiment'] for _,r in copies})==1:
            available[copies[0][1]['airline_sentiment']].append(copies[0])
    rng=random.Random(20260924);splits={k:[] for k in ['train','dev','calibration','test']}
    for label,n in zip(OPTIONS,[134,133,133]):
        chosen=rng.sample(available[label],400+2*n)
        for name,lo,hi in [('train',0,300),('dev',300,400),('calibration',400,400+n),('test',400+n,400+2*n)]:
            splits[name].extend(chosen[lo:hi])
    output.mkdir(parents=True,exist_ok=False)
    manifest=dict(seed=20260924,source=SOURCE,archive_sha256=ARCHIVE_SHA256,excluded_prior_unique=len(excluded),
        labels={},splits={},protocol=dict(model='gemma-4-E2B-it-Q8_0.gguf',
        model_sha256='996d08777aadc6bfd3c7375ef70ba25a0f55240075860754fdb18d6d860aa63a',
        methods=['base_temperature','logit_affine','hidden'],l2_grid=[.001,.01,.1,1.],
        objective='mean cross entropy + L2/2 * squared standardized weights; CPU float64 LBFGS max_iter=200',
        standardization='train only; std clamp 1e-6; fold into exported weights',
        hyperparameter_selection='lowest dev NLL before temperature; never test',
        calibration='temperature fit only on independent 400 calibration cases',
        deployment_selection='lowest dev NLL across trained heads; explicit opt-in',
        scope='Balanced task-specific head. Normalized exact dedup, not semantic dedup. Not RLCD. Base candidate mass gate preserved.'))
    records={}
    for name,selected in splits.items():
        rng.shuffle(selected);records[name]=[]
        for rot in (range(3) if name=='train' else range(1)):
            for i,r in selected:
                cid=f'airline-row-{i:05d}-r{rot}';opts=list(OPTIONS);opts=opts[rot:]+opts[:rot]
                records[name].append(dict(id=cid,request=dict(state={'tweet':r['text']},decisions=[dict(
                    id='airline_sentiment',instruction=INSTRUCTION,kind=dict(type='choice',options=[dict(id=k,criterion=OPTIONS[k]) for k in opts]))])))
                manifest['labels'][cid]=r['airline_sentiment']
    records['probe']=[]
    for label in OPTIONS:
        for case in [c for c in records['test'] if manifest['labels'][c['id']]==label][:20]:
            for rot in range(3):
                c=copy.deepcopy(case);c['id']=case['id'].removesuffix('-r0')+f'-probe-r{rot}'
                opts=c['request']['decisions'][0]['kind']['options'];c['request']['decisions'][0]['kind']['options']=opts[rot:]+opts[:rot]
                records['probe'].append(c);manifest['labels'][c['id']]=label
    for name,items in records.items():
        path=output/(name+'.jsonl');path.write_text(''.join(json.dumps(c,ensure_ascii=False)+'\n' for c in items))
        manifest['splits'][name]=dict(ids=[c['id'] for c in items],sha256=sha256(path))
    assert all({identity(r['text']) for _,r in splits[a]}.isdisjoint(identity(r['text']) for _,r in splits[b])
               for a in splits for b in splits if a!=b)
    (output/'selection.json').write_text(json.dumps(manifest,indent=2)+'\n')
    print({k:len(v) for k,v in records.items()})

if __name__=='__main__':
    p=argparse.ArgumentParser();p.add_argument('--source',type=Path,required=True);p.add_argument('--previous',type=Path,required=True);p.add_argument('--output',type=Path,required=True)
    a=p.parse_args();prepare(a.source,a.previous,a.output)

File: crates/l2s1-tools/src/prepare_output_head.rs
use crate::{
    common::{read_jsonl, sha256, write_json},
    kaggle_airline::{ARCHIVE_SHA256, LABELS, Row, available, case_json, identity, read_archive},
    python_random::PythonRandom,
};
use anyhow::{Context, Result, ensure};
use clap::Args as ClapArgs;
use serde_json::json;
use std::{
    collections::{BTreeMap, HashSet},
    fs::{self, File},
    io::Write,
    path::PathBuf,
};

#[derive(ClapArgs)]
pub struct Args {
    #[arg(long)]
    source: PathBuf,
    #[arg(long)]
    previous: PathBuf,
    #[arg(long)]
    output: PathBuf,
}

pub fn run(args: Args) -> Result<()> {
    let mut excluded = HashSet::new();
    for (folder, names) in [
        (&args.source, vec!["fit", "validation"]),
        (&args.previous, vec!["train", "calibration", "test"]),
    ] {
        for name in names {
            for row in read_jsonl(&folder.join(format!("{name}.jsonl")))? {
                excluded.insert(identity(
                    row["request"]["state"]["tweet"]
                        .as_str()
                        .context("missing prior tweet")?,
                ));
            }
        }
    }
    let rows = read_archive(&args.source.join("dataset.zip"))?;
    let (groups, _) = available(&rows, &excluded)?;
    let mut rng = PythonRandom::new(20260924);
    let mut splits: BTreeMap<&str, Vec<(usize, Row)>> = [
        ("train", Vec::new()),
        ("dev", Vec::new()),
        ("calibration", Vec::new()),
        ("test", Vec::new()),
    ]
    .into();
    for (group, n) in groups.iter().zip([134, 133, 133]) {
        ensure!(group.len() >= 400 + 2 * n, "too few eligible airline rows");
        let sampled = rng.sample_indices(group.len(), 400 + 2 * n);
        for (name, lo, hi) in [
            ("train", 0, 300),
            ("dev", 300, 400),
            ("calibration", 400, 400 + n),
            ("test", 400 + n, 400 + 2 * n),
        ] {
            splits
                .get_mut(name)
                .unwrap()
                .extend(sampled[lo..hi].iter().map(|i| group[*i].clone()));
        }
    }
    fs::create_dir(&args.output)?;
    let mut labels = serde_json::Map::new();
    let mut meta = serde_json::Map::new();
    for name in ["train", "dev", "calibration", "test"] {
        let selected = splits.get_mut(name).unwrap();
        rng.shuffle(selected);
        let path = args.output.join(format!("{name}.jsonl"));
        let mut file = File::create(&path)?;
        let mut ids = Vec::new();
        for rotation in 0..if name == "train" { 3 } else { 1 } {
            for (index, row) in selected.iter() {
                let id = format!("airline-row-{index:05}-r{rotation}");
                writeln!(file, "{}", case_json(&id, &row.text, rotation)?)?;
                labels.insert(id.clone(), json!(row.sentiment));
                ids.push(id);
            }
        }
        file.flush()?;
        meta.insert(name.to_owned(), json!({"ids":ids,"sha256":sha256(&path)?}));
    }
    let mut probe = File::create(args.output.join("probe.jsonl"))?;
    let mut probe_ids = Vec::new();
    for label in LABELS {
        for (index, row) in splits["test"]
            .iter()
            .filter(|(_, r)| r.sentiment == label)
            .take(20)
        {
            for rotation in 0..3 {
                let id = format!("airline-row-{index:05}-probe-r{rotation}");
                writeln!(probe, "{}", case_json(&id, &row.text, rotation)?)?;
                labels.insert(id.clone(), json!(label));
                probe_ids.push(id);
            }
        }
    }
    probe.flush()?;
    meta.insert(
        "probe".to_owned(),
        json!({"ids":probe_ids,"sha256":sha256(&args.output.join("probe.jsonl"))?}),
    );
    let unique = |name: &str| {
        splits[name]
            .iter()
            .map(|(_, r)| identity(&r.text))
            .collect::<HashSet<_>>()
    };
    for (i, a) in ["train", "dev", "calibration", "test"].iter().enumerate() {
        for b in ["train", "dev", "calibration", "test"].iter().skip(i + 1) {
            ensure!(unique(a).is_disjoint(&unique(b)), "split texts overlap");
        }
    }
    let manifest = json!({"seed":20260924,"source":"https://www.kaggle.com/datasets/crowdflower/twitter-airline-sentiment",
        "archive_sha256":ARCHIVE_SHA256,"excluded_prior_unique":excluded.len(),"labels":labels,"splits":meta,
        "protocol":{"model":"gemma-4-E2B-it-Q8_0.gguf","model_sha256":"996d08777aadc6bfd3c7375ef70ba25a0f55240075860754fdb18d6d860aa63a",
            "methods":["base_temperature","logit_affine","hidden"],"l2_grid":[0.001,0.01,0.1,1.0],
            "objective":"mean cross entropy + L2/2 * squared standardized weights; CPU float64 LBFGS max_iter=200",
            "standardization":"train only; std clamp 1e-6; fold into exported weights",
            "hyperparameter_selection":"lowest dev NLL before temperature; never test",
            "calibration":"temperature fit only on independent 400 calibration cases",
            "deployment_selection":"lowest dev NLL across trained heads; explicit opt-in",
            "scope":"Balanced task-specific head. Normalized exact dedup, not semantic dedup. Not RLCD. Base candidate mass gate preserved."}});
    write_json(&args.output.join("selection.json"), &manifest)?;
    println!("{}", serde_json::to_string_pretty(&manifest["splits"])?);
    Ok(())
}
